File: wordpress_publisher.py
from wordpress_xmlrpc import Client, WordPressPost
from wordpress_xmlrpc.methods.posts import NewPost
from wordpress_xmlrpc.compat import xmlrpc_client
from PIL import Image
import io
import os
import requests

def post_to_wordpress(article):
    client = Client(
        os.getenv('WORDPRESS_HOST'),
        os.getenv('WORDPRESS_USERNAME'),
        os.getenv('WORDPRESS_PASSWORD')
    )

    post = WordPressPost()
    post.title = article['title']
    post.content = format_content(article)
    post.post_status = 'publish'
    post.terms_names = {
        'post_tag': article['keywords'],
        'category': ['Generated Content']  # You might want to adjust this
    }

    # Set the language if your WordPress setup supports it
    if 'language' in article:
        post.custom_fields = [{'key': 'lang', 'value': article['language']}]

    # Upload and set featured image if available
    if 'image' in article:
        image_id = upload_featured_image(client, article['image'], article['title'])
        if image_id:
            post.thumbnail = image_id

    post_id = client.call(NewPost(post))
    print(f"Article published successfully. Post ID: {post_id}")

def format_content(article):
    # You can implement more sophisticated formatting here
    return article['content']

def upload_featured_image(client, image_url, title):
    response = requests.get(image_url)
    if response.status_code != 200:
        print(f"Failed to download image from {image_url}")
        return None

    # Process and optimize the image
    img = Image.open(io.BytesIO(response.content))
    img.thumbnail((1200, 630))  # Resize image
    img_byte_arr = io.BytesIO()
    img.save(img_byte_arr, format='JPEG')
    img_data = img_byte_arr.getvalue()

    data = {
        'name': f'{title}.jpg',
        'type': 'image/jpeg',
        'bits': xmlrpc_client.Binary(img_data)
    }

    response = client.call(xmlrpc_client.methodcall('wp.uploadFile', [0, os.getenv('WORDPRESS_USERNAME'), os.getenv('WORDPRESS_PASSWORD'), data]))
    
    if 'id' in response:
        return response['id']
    else:
        print("Failed to upload image")
        return None

File: src/wordpressPublisher.js
import xmlrpc from 'xmlrpc';
import fs from 'fs/promises';
import path from 'path';
import sharp from 'sharp';

export async function postToWordPress(article) {
  const client = xmlrpc.createSecureClient({
    host: process.env.WORDPRESS_HOST,
    port: 443,
    path: '/xmlrpc.php',
  });

  const username = process.env.WORDPRESS_USERNAME;
  const password = process.env.WORDPRESS_PASSWORD;

  try {
    const categories = await getCategories(client, username, password);
    const categoryId = findRelevantCategory(categories, article.keywords);

    let imageUrl = null;
    if (article.type === 'web' && article.contents.some(content => content.image)) {
      imageUrl = await uploadFeaturedImage(client, username, password, article.contents[0].image, article.title);
    }

    const post = {
      post_type: 'post',
      post_title: article.title,
      post_content: article.content,
      post_status: 'publish',
      terms: {
        category: [categoryId],
        post_tag: article.keywords,
      },
      post_thumbnail: imageUrl,
    };

    if (article.language) {
      post.post_language = article.language;
    }

    const postId = await new Promise((resolve, reject) => {
      client.methodCall('wp.newPost', [0, username, password, post], (error, value) => {
        if (error) reject(error);
        else resolve(value);
      });
    });

    console.log(`Article published successfully in ${article.language}. Post ID: ${postId}`);
  } catch (error) {
    console.error('Error publishing to WordPress:', error);
  }
}

async function getCategories(client, username, password) {
  return new Promise((resolve, reject) => {
    client.methodCall('wp.getTerms', [0, username, password, 'category'], (error, categories) => {
      if (error) reject(error);
      else resolve(categories);
    });
  });
}

function findRelevantCategory(categories, keywords) {
  // Implementăm logica de găsire a categoriei relevante bazată pe cuvintele cheie
  // Pentru simplitate, vom returna prima categorie găsită
  return categories[0].term_id;
}

async function uploadFeaturedImage(client, username, password, imageUrl, title) {
  const imageData = await fetch(imageUrl).then(res => res.buffer());
  const optimizedImageData = await sharp(imageData)
    .resize(1200, 630, { fit: 'inside' })
    .toBuffer();

  const base64Image = optimizedImageData.toString('base64');

  const uploadedImage = await new Promise((resolve, reject) => {
    client.methodCall('wp.uploadFile', [
      0,
      username,
      password,
      {
        name: `${title}.jpg`,
        type: 'image/jpeg',
        bits: base64Image,
      },
    ], (error, response) => {
      if (error) reject(error);
      else resolve(response);
    });
  });

  return uploadedImage.url;
}